docs/adr/0005-swift-dispatchqueue-usage.md

```diff
@@ -0,0 +1,73 @@
+
+# Remove Unnecessary Swift `DispatchQueue` Usage
+
+## Status
+
+* Proposed
+
+## Context
+
+### `DispatchQueue` Overview
+
+A `DispatchQueue` is a Swift abstraction used to manage the serial or concurrent execution of tasks. When a `DispatchQueue` is created, a thread pool (comprised of one or more reallocated or newly-created threads) is assigned to the queue by the system. `DispatchQueue`s can be configured to execute tasks, or "work items", serially by default or concurrently if the [attributes](https://developer.apple.com/documentation/dispatch/dispatchqueue/attributes) parameter of the constructor is appropriately set.
+
+Other [constructor parameters](https://developer.apple.com/documentation/dispatch/dispatchqueue/2300059-init) allow for additional configurations but they are not relevant here as the custom queues created in our rust components and in the corresponding [firefox-ios storage layer](https://github.com/mozilla-mobile/firefox-ios/tree/main/Storage/Rust) set the `label` parameter of the `DispatchQueue` constructor at most and therefore execute serially with default settings.
+
+We also make use of the two non-custom `DispatchQueue`s, the [main](https://developer.apple.com/documentation/dispatch/dispatchqueue/1781006-main) and [global](https://developer.apple.com/documentation/dispatch/dispatchqueue/2300077-global) queues, in the FxA component. Unsurprisingly, the main queue executes on the main thread. The global queues (high, default, low, and background queues) execute concurrently and are shared by the whole system. Generally, best practice dictates that the main queue be reserved for UI/UX updates and global and custom queues (which are ultimately performed by global queues) do the remainder of the work.
+
+Lastly, tasks can be submitted to a `DispatchQueue` either synchronously (which blocks the queue until the submitted task is complete) or asynchronously (which returns immediately after the task is submitted to the queue). Best practice advises that synchronous calls be used only when needed and should be avoided entirely when scheduleding tasks concurrently or on the main queue.
+
+### Our `DispatchQueue` Usage
+
+There are places in application services code where `DispatchQueues` are appropriately used. In the FxA component where we have functions that update the UI (mostly via [closure function parameters](https://github.com/mozilla/application-services/blob/451bcc2fe8fe6675ca3de962a4b38b3fa181a806/components/fxa-client/ios/FxAClient/FxAccountManager.swift#L108)) or [dispatch notifications via `NotificationCenter`](https://github.com/mozilla/application-services/blob/72b827c3e0f883163762857fd766df1aeb060725/components/fxa-client/ios/FxAClient/FxAccountDeviceConstellation.swift#L49), our use of non-custom queues makes sense.
```

Comment:
I'm mildly surprised that is the right thing to do for something like app-services, and the appropriate async policy can't still be owned by the main app (eg, it's not obvious from reading that FxA code that the completion handler has anything to do with UI at all). However, this is way above my pay-grade, so if this is the opinion of the iOS team it's fine with me.

Reply:
@mhammond I was squinting when I made this call so I'd be very open to being wrong because this is _way_ above my pay-grade as well. I also read that sending Notification Center events doesn't require queues but I can't find that blog post for the life of me.

The long and short of it is I'm way more confident in my assessment of the storage layer of the logins, bookmarks, and tabs component than I am about our usage of queues in FxA or Rust_Log.

Reply:
@nbhasin2 @OrlaM Would you agree with my assessment of the `DispatchQueue` usage in FxA and RustLog?

Reply:
It looks like that callback goes to a file that updates a web view within FxA (FxAWebViewController) which is UI so it probably does need to be on the main thread, but in this scenario it would be more clear if FxAWebViewController ensured it was on the main thread when doing updates instead of the business logic needing to be aware of how it's to be used.

For the changes to the DispatchQueue usage I'm having a hard time telling whats actually happening when a rust call is being made. Is it making a HTTP request, updating a data base or a combination? Is thread safety handled at a lower level and therefore it's redundant at this level?

If it's making a database update then the DispatchQueue calls made sense. This is currently the best way in Swift of making the database calls thread safe. But if this is already being handled at a lower level then these calls would be redundant. 

If it's making a HTTP request then it doesn't necessarily need to be handled by a custom queue because serialization doesn't matter but it does need to be moved off the main thread in some fashion so it doesn't block the UI.

If the concern is the duplication of this logic between LoginsStorage.swift and RustLogins.swift then it's better to remove it from RustLogins.swift instead. It's better for these kinds of thread safety checks to happen at the lowest possible level in my opinion.